File: javascript/calcium.js
var CalciumSprite = function(x, y, frames, frame_index) {
	this.x = x;
	this.y = y;
	this.frames = frames;
	this.frame_index = frame_index || 0;
};

CalciumSprite.prototype.get_pixels = function () {
	return this.frames[this.frame_index];
};

CalciumSprite.prototype.next_frame = function () {
	this.frame_index += 1;
	if (this.frame_index >= this.frames.length) {
		this.frame_index = 0;
	}
};

CalciumSprite.prototype.last_frame = function () {
	this.frame_index -= 1;
	if (this.frame_index < 0) {
		this.frame_index = this.frames.length - 1;
	}
};

var CalciumScreen = function(width, height, fps) {
	this.width = width;
	this.height = height || width;
	this.fps = fps || 60;
	this.clear_color = 3;
	this.lines = [];
	this.clear();
};

CalciumScreen.LEVELS = [ '█', '▓', '▒', '░', ' ' ];

CalciumScreen.prototype.clear = function () {
	this.lines = [];
	var line;
	for (var li=0; li < this.height; li++) {
		line = [];
		for (var ci=0; ci < this.width; ci++) {
			line.push(CalciumScreen.LEVELS[this.clear_color]);
		}
		this.lines.push(line);
	}
};

CalciumScreen.prototype.toString = function () {
	var r = '';
	for (var li=0; li < this.height; li++) {
		for (var ci=0; ci < this.width; ci++) {
			r += this.lines[li][ci] + this.lines[li][ci];
		}
		r += '\n';
	}
	return r;
};

CalciumScreen.prototype.set_pixel = function (x, y, level) {
	if (x < 0 || x >= this.width) {
		return;
	}
	if (y < 0 || y >= this.height) {
		return;
	}
	if (level > 4) {
		console.warn('Calcium: invalid level value');
		return;
	}
	this.lines[y][x] = CalciumScreen.LEVELS[level];
};

CalciumScreen.prototype.plot = function (sprite) {
	var pixels = sprite.get_pixels();
	var length = pixels.length;
	for (var pi = 0; pi < length; pi += 3) {
		this.set_pixel(
			sprite.x + pixels[pi],
			sprite.y + pixels[pi + 1],
			pixels[pi + 2]
		);
	}
};

CalciumScreen.prototype.mainloop = function (func) {
	setInterval(function() {
		func();
	}, 1000 / this.fps);
};

CalciumScreen.prototype.draw = function (elem) {
	elem.innerHTML = this.toString();
};

File: python/calcium.py
# coding: utf-8

import copy


class CalciumSprite(object):
    def __init__(self, x, y, animations, frame_index=0, animation_key=None):
        self.x = x
        self.y = y
        self.animations = animations
        self.animation_key = animation_key or list(animations.keys())[0]
        self.frame_index = frame_index

    def get_pixels(self):
        return self.animations.get(self.animation_key)[self.frame_index]

    def next_frame(self):
        self.frame_index += 1
        if self.frame_index >= len(self.animations.get(self.animation_key)):
            self.frame_index = 0

    def last_frame(self):
        self.frame_index -= 1
        if self.frame_index < 0:
            self.frame_index = len(self.animations.get(self.animation_key)) - 1

    def clone(self):
        return copy.deepcopy(self)

    def invert_colors(self):
        pixels = self.get_pixels()
        for i in range(0, len(pixels), 3):
            pixels[i + 2] = 3 - pixels[i + 2]

    def dark(self, level=1):
        pixels = self.get_pixels()
        for i in range(0, len(pixels), 3):
            pixels[i + 2] = max(0, pixels[i + 2] - level)

    def light(self, level=1):
        pixels = self.get_pixels()
        for i in range(0, len(pixels), 3):
            pixels[i + 2] = min(pixels[i + 2] + level, 3)


class CalciumScreen(object):
    LEVELS = ['█', '▓', '▒', '░']

    def __init__(self, width, height=None):
        u"""Represent a virtual screen."""
        self.width = width
        self.height = height or width
        self.clear_color = 3
        self.lines = []
        self.clear()

    def clear(self):
        self.lines = []
        for li in range(self.height):
            line = []
            for ci in range(self.width):
                line.append(CalciumScreen.LEVELS[self.clear_color])
            self.lines.append(line)

    def __repr__(self):
        return '\n'.join([''.join([c + c for c in line]) for line in self.lines])

    def set_pixel(self, x, y, level):
        if x < 0 or x >= self.width:
            return
        if y < 0 or y >= self.height:
            return
        if level > 4:
            print('Calcium: invalid level value')
            return
        self.lines[int(y)][int(x)] = CalciumScreen.LEVELS[level]

    def plot(self, sprite):
        pixels = sprite.get_pixels()
        for i in range(0, len(pixels), 3):
            self.set_pixel(
                sprite.x + pixels[i],
                sprite.y + pixels[i + 1],
                pixels[i + 2]
            )